File: src/core/Stocks/StockList.cs
using System;
using System.Collections.Generic;
using core.Shared;

namespace core.Stocks
{
    public class StockListState : IAggregateState
    {
        public Guid Id { get; private set; }
        public Guid UserId { get; private set; }
        public string Name { get; private set; }
        public string Description { get; private set; }
        public List<StockListTicker> Tickers { get; } = new();
        public HashSet<string> Tags { get; } = new();

        public void Apply(AggregateEvent e) => ApplyInternal(e);

        public bool ContainsTag(string tag) => Tags.Contains(tag);

        protected void ApplyInternal(dynamic obj) => ApplyInternal(obj);

        private void ApplyInternal(StockListCreated created)
        {
            Description = created.Description;
            Id = created.AggregateId;
            Name = created.Name;
            UserId = created.UserId;
        }

        private void ApplyInternal(StockListUpdated updated)
        {
            Description = updated.Description;
            Name = updated.Name;
        }

        private void ApplyInternal(StockListTickerAdded added) =>
            Tickers.Add(new StockListTicker(added.Note, new Ticker(added.Ticker), added.When));

        private void ApplyInternal(StockListTickerRemoved removed) =>
            Tickers.RemoveAll(x => x.Ticker.Value == removed.Ticker);

        private void ApplyInternal(StockListTagAdded added) =>
            Tags.Add(added.Tag);

        private void ApplyInternal(StockListTagRemoved removed) =>
            Tags.Remove(removed.Tag);
        
        private void ApplyInternal(StockListCleared cleared) =>
            Tickers.Clear();
    }

    public record StockListTicker(string Note, Ticker Ticker, DateTimeOffset When);
    
    public class StockList : Aggregate<StockListState>
    {
        public StockList(IEnumerable<AggregateEvent> events) : base(events)
        {
        }

        public StockList(string description, string name, Guid userId)
        {
            if (userId == Guid.Empty)
            {
                throw new InvalidOperationException("Missing user id");
            }

            if (string.IsNullOrWhiteSpace(name))
            {
                throw new InvalidOperationException("Missing list name");
            }

            Apply(new StockListCreated(Guid.NewGuid(), Guid.NewGuid(), DateTimeOffset.UtcNow, description, name, userId));
        }

        public void Update(string name, string description)
        {
            if (string.IsNullOrWhiteSpace(name))
            {
                throw new InvalidOperationException("Missing name");
            }

            if (State.Name == name && State.Description == description)
            {
                return;
            }

            Apply(new StockListUpdated(Guid.NewGuid(), State.Id, DateTimeOffset.UtcNow, description, name));
        }

        public void AddStock(Ticker ticker, string note)
        {
            var exists = State.Tickers.Exists(x => x.Ticker.Equals(ticker));
            if (exists)
            {
                return;
            }
            
            Apply(new StockListTickerAdded(Guid.NewGuid(), State.Id, DateTimeOffset.UtcNow, note, ticker.Value));
        }

        public void RemoveStock(Ticker ticker)
        {
            var exists = State.Tickers.Exists(x => x.Ticker.Equals(ticker));
            if (!exists)
            {
                throw new InvalidOperationException("Ticker does not exist in the list");
            }

            Apply(new StockListTickerRemoved(Guid.NewGuid(), State.Id, DateTimeOffset.UtcNow, ticker.Value));
        }

        public void AddTag(string tag)
        {
            if (string.IsNullOrWhiteSpace(tag))
            {
                throw new InvalidOperationException("Missing tag");
            }

            if (State.ContainsTag(tag))
            {
                return;
            }

            Apply(new StockListTagAdded(Guid.NewGuid(), State.Id, DateTimeOffset.UtcNow, tag));
        }

        public void RemoveTag(string tag)
        {
            if (string.IsNullOrWhiteSpace(tag))
            {
                throw new InvalidOperationException("Missing tag");
            }

            if (!State.ContainsTag(tag))
            {
                return;
            }

            Apply(new StockListTagRemoved(Guid.NewGuid(), State.Id, DateTimeOffset.UtcNow, tag));
        }
        
        public void Clear()
        {
            if (State.Tickers.Count == 0)
            {
                return;
            }

            Apply(new StockListCleared(Guid.NewGuid(), State.Id, DateTimeOffset.UtcNow));
        }
    }
}

File: src/core2.fs/Stocks/StockList.fs
namespace core.Stocks

open System
open System.Collections.Generic
open core.Shared

[<Struct>]
type StockListTicker =
    { Note: string
      Ticker: Ticker
      When: DateTimeOffset }

type StockListState() =
    let mutable _id = Guid.Empty
    let mutable _userId = Guid.Empty
    let mutable _name: string = null
    let mutable _description: string = null
    let _tickers = List<StockListTicker>()
    let _tags = HashSet<string>()

    member _.Id = _id
    member _.UserId = _userId
    member _.Name = _name
    member _.Description = _description
    member _.Tickers = _tickers :> IList<_>
    member _.Tags = _tags :> HashSet<_>

    member this.Apply(e: AggregateEvent) =
        this.ApplyInternal(e :> obj)

    member _.ContainsTag(tag: string) = _tags.Contains(tag)

    member private this.ApplyInternal(obj: obj) =
        match obj with
        | :? StockListCreated as e -> this.ApplyInternal(e)
        | :? StockListUpdated as e -> this.ApplyInternal(e)
        | :? StockListTickerAdded as e -> this.ApplyInternal(e)
        | :? StockListTickerRemoved as e -> this.ApplyInternal(e)
        | :? StockListTagAdded as e -> this.ApplyInternal(e)
        | :? StockListTagRemoved as e -> this.ApplyInternal(e)
        | :? StockListCleared as e -> this.ApplyInternal(e)
        | _ -> ()

    member private this.ApplyInternal(created: StockListCreated) =
        _description <- created.Description
        _id <- created.AggregateId
        _name <- created.Name
        _userId <- created.UserId

    member private this.ApplyInternal(updated: StockListUpdated) =
        _description <- updated.Description
        _name <- updated.Name

    member private this.ApplyInternal(added: StockListTickerAdded) =
        _tickers.Add({ Note = added.Note; Ticker = Ticker(added.Ticker); When = added.When })

    member private this.ApplyInternal(removed: StockListTickerRemoved) =
        _tickers.RemoveAll(fun x -> x.Ticker.Value = removed.Ticker) |> ignore

    member private this.ApplyInternal(added: StockListTagAdded) =
        _tags.Add(added.Tag) |> ignore

    member private this.ApplyInternal(removed: StockListTagRemoved) =
        _tags.Remove(removed.Tag) |> ignore

    member private this.ApplyInternal(cleared: StockListCleared) =
        _tickers.Clear()

    interface IAggregateState with
        member this.Id = this.Id
        member this.Apply(e) = this.Apply(e)

type StockList =
    inherit Aggregate<StockListState>

    new (events: IEnumerable<AggregateEvent>) = { inherit Aggregate<StockListState>(events) }

    new (description: string, name: string, userId: Guid) as this =
        { inherit Aggregate<StockListState>() }
        then
            if userId = Guid.Empty then
                raise (InvalidOperationException("Missing user id"))

            if String.IsNullOrWhiteSpace(name) then
                raise (InvalidOperationException("Missing list name"))

            this.Apply(StockListCreated(Guid.NewGuid(), Guid.NewGuid(), DateTimeOffset.UtcNow, description, name, userId))

    member this.Update(name: string, description: string) =
        if String.IsNullOrWhiteSpace(name) then
            raise (InvalidOperationException("Missing name"))

        if this.State.Name = name && this.State.Description = description then
            ()
        else
            this.Apply(StockListUpdated(Guid.NewGuid(), this.State.Id, DateTimeOffset.UtcNow, description, name))

    member this.AddStock(ticker: Ticker, note: string) =
        let exists = this.State.Tickers |> Seq.exists (fun x -> x.Ticker.Equals(ticker))
        if not exists then
            this.Apply(StockListTickerAdded(Guid.NewGuid(), this.State.Id, DateTimeOffset.UtcNow, note, ticker.Value))

    member this.RemoveStock(ticker: Ticker) =
        let exists = this.State.Tickers |> Seq.exists (fun x -> x.Ticker.Equals(ticker))
        if not exists then
            raise (InvalidOperationException("Ticker does not exist in the list"))

        this.Apply(StockListTickerRemoved(Guid.NewGuid(), this.State.Id, DateTimeOffset.UtcNow, ticker.Value))

    member this.AddTag(tag: string) =
        if String.IsNullOrWhiteSpace(tag) then
            raise (InvalidOperationException("Missing tag"))

        if not (this.State.ContainsTag(tag)) then
            this.Apply(StockListTagAdded(Guid.NewGuid(), this.State.Id, DateTimeOffset.UtcNow, tag))

    member this.RemoveTag(tag: string) =
        if String.IsNullOrWhiteSpace(tag) then
            raise (InvalidOperationException("Missing tag"))

        if this.State.ContainsTag(tag) then
            this.Apply(StockListTagRemoved(Guid.NewGuid(), this.State.Id, DateTimeOffset.UtcNow, tag))

    member this.Clear() =
        if this.State.Tickers.Count > 0 then
            this.Apply(StockListCleared(Guid.NewGuid(), this.State.Id, DateTimeOffset.UtcNow))
